csi/controller_server.go

```diff
@@ -545,6 +546,34 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 		return nil, status.Error(codes.InvalidArgument, "Snapshot name must be provided")
 	}
 
+	nodeID, err := cs.autoAttachVolumeToRandomNode(csiVolumeName)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if nodeID != "" {
+		defer func() {
+			volume, err := cs.apiClient.Volume.ById(csiVolumeName)
+			if err != nil {
+				logrus.WithError(err).Errorf("volume %s unavailable", csiVolumeName)
+			}
+			if volume == nil {
+				logrus.WithError(err).Errorf("volume %s not found", csiVolumeName)
+			}
+			logrus.Debugf("CSI snapshot attempting to unattach %s volume on %s node", csiVolumeName, nodeID)
+			cs.unpublishVolume(volume, nodeID, func() error {
```

Comment:
If somehow, the volume was not unpublished successfully. How is it? What users will encounter?

Reply:
If the volume is already attached before creating a snapshot, the volume will be automatically detached after snapshot?

Reply:
AFAIK CSI snapshot can only have PVC as source where volume should have been already attached. Hense this fix is only for the csi-sanity test.

Ref: https://kubernetes.io/docs/concepts/storage/volume-snapshots/#volumesnapshots

Reply:
> If the volume is already attached before creating a snapshot, the volume will be automatically detached after snapshotting?

It's been handled by
```
	nodeID, err := cs.autoAttachVolumeToRandomNode(csiVolumeName)
	if err != nil {
		return nil, status.Error(codes.Internal, err.Error())
	}
	if nodeID != "" {
```

Reply:
> AFAIK CSI snapshot can only have PVC as source where volume should have been already attached.
A Longhorn volume that has a corresponding PVC can be detached.

Reply:
And I think the CSI plugin should check if the volume is already attached before picking up a random node (then start auto attachment).

Reply:
It's handled in autoAttachVolumeToRandomNode.
```
func (cs *ControllerServer) autoAttachVolumeToRandomNode(volumeName string) (nodeID string, err error) {
	volume, err := cs.apiClient.Volume.ById(volumeName)
	if err != nil {
		return "", errors.Wrapf(err, "failed to get volume %v", volumeName)
	}
	if volume.State == string(longhorn.VolumeStateDetached) {
```

Reply:
Oh, OK. I misread it... But can you separate the check from this function? It's like we will blindly do auto attachment for the CSI snapshot creation at the 1st glance.

Reply:
OK!